alica_engine/src/engine/RuntimePlanFactory.cpp

```diff
@@ -22,11 +22,23 @@ RuntimePlanFactory::RuntimePlanFactory(
 
 void RuntimePlanFactory::init(std::unique_ptr<IPlanCreator>&& pc)
 {
-    _creator = std::move(pc);
+    {
+        std::lock_guard lk(_m);
+        _creator = std::move(pc);
+        _initialized = true;
+    }
+    _cv.notify_all();
 }
 
 std::unique_ptr<BasicPlan> RuntimePlanFactory::create(int64_t id, const Plan* planModel) const
 {
+    {
+        std::unique_lock lk(_m);
+        while (!_initialized) {
+            _cv.wait(lk);
+        }
+    }
+
     PlanContext ctx{_globalBlackboard, planModel->getName(), planModel, _traceFactory};
     std::unique_ptr<BasicPlan> basicPlan = _creator->createPlan(id, ctx);
```

Comment:
Should `_creator` usage also be under the lock still?

Reply:
Initially it was called as-is. And seems inside we have only creation routines.
Strongly speaking I keep old behaviour here (except the entering lock).

Reply:
> Just from the memory correctness perspective i think it should be under the lock if i don't miss anything.

Maybe, but it should be analyzed to make a final decision. It's like if we are only reading something which already was propagated it should be ok.